src/main/java/net/logstash/logback/util/ThreadLocalHolder.java

```diff
@@ -148,39 +149,42 @@ public void close() {
          * immediately instead of being recycled.
          */
         this.closed = true;
-        
+
         /*
          * Dispose value assigned to running threads.
          * "inuse" values will be disposed by the owning thread when it releases it.
          */
-        for (HolderRef holderRef: this.threadValues) {
+        for (HolderRef holderRef: this.threadValues.values()) {
             Holder<T> holder = holderRef.getHolder();
             if (!holder.leased) {
                 disposeHolder(holder);
             }
         }
         this.threadValues.clear();
-     
+
         /*
          * Dispose values assigned to threads that just died
          */
         processDeadThreads();
     }
-    
-    
+
+
     /**
      * Create a new {@link Holder} and keep track of the asking thread for clearing when the thread
      * is gone.
-     * 
+     *
      * @return a {@link Holder} assigned to the current thread.
      */
     private Holder<T> initializeThread() {
-        Holder<T> holder = new Holder<>();
-        threadValues.add(new HolderRef(Thread.currentThread(), holder, deadThreads));
-        return holder;
+        final Thread currentThread = Thread.currentThread();
+        final long threadId = currentThread.getId();
+        return threadValues.computeIfAbsent(
```

Comment:
I would add a comment here about the reason why the value associated with the current thread is stored in the threadValues Map and not in the ThreadLocal itself as it was before. Making reference to the ForkJoinPool issue clearing the thread local while the thread is still alive would be nice as well.

Note that this strategy makes the assumption that Thread.getId() does not change after a thread is created. This seems to be confirmed by the javadoc: the id seems to be immutable and unique...

Reply:
Also, you need to rebase to pickup the latest changes otherwise the build will keep failing... (not related to your PR)

Reply:
added